packages/protocol/contracts/layer1/based/TaikoInbox.sol

```diff
@@ -0,0 +1,723 @@
+
+                emit Stats1Updated(stats1);
+
+                // Ask signal service to write cross chain signal
+                ISignalService(resolve(LibStrings.B_SIGNAL_SERVICE, false)).syncChainData(
+                    _config.chainId, LibStrings.H_STATE_ROOT, synced.blockId, synced.stateRoot
+                );
+            }
+        }
+
+        state.stats2 = _stats2;
+        emit Stats2Updated(_stats2);
+    }
+
+    function _debitBond(address _user, uint256 _amount) private {
+        if (_amount == 0) return;
+
+        uint256 balance = state.bondBalance[_user];
+        if (balance >= _amount) {
+            unchecked {
+                state.bondBalance[_user] = balance - _amount;
+            }
+        } else {
+            _handleDeposit(_user, _amount);
+        }
+        emit BondDebited(_user, _amount);
+    }
+
+    function _creditBond(address _user, uint256 _amount) private {
+        if (_amount == 0) return;
+        unchecked {
+            state.bondBalance[_user] += _amount;
+        }
+        emit BondCredited(_user, _amount);
+    }
+
+    function _handleDeposit(address _user, uint256 _amount) private {
+        address bond = bondToken();
+
+        if (bond != address(0)) {
+            require(msg.value == 0, MsgValueNotZero());
+            IERC20(bond).transferFrom(_user, address(this), _amount);
+        } else {
+            require(msg.value == _amount, EtherNotPaidAsBond());
+        }
+        emit BondDeposited(_user, _amount);
+    }
+
+    function _validateBatchParams(
+        BatchParams memory _params,
+        uint64 _maxAnchorHeightOffset,
+        uint8 _maxSignalsToReceive,
+        uint16 _maxBlocksPerBatch,
+        Batch memory _lastBatch
+    )
+        private
+        view
+        returns (uint64 anchorBlockId_, uint64 lastBlockTimestamp_)
+    {
+        unchecked {
+            if (_params.anchorBlockId == 0) {
+                anchorBlockId_ = uint64(block.number - 1);
+            } else {
+                require(
+                    _params.anchorBlockId + _maxAnchorHeightOffset >= block.number,
+                    AnchorBlockIdTooSmall()
+                );
+                require(_params.anchorBlockId < block.number, AnchorBlockIdTooLarge());
+                require(
+                    _params.anchorBlockId >= _lastBatch.anchorBlockId,
+                    AnchorBlockIdSmallerThanParent()
+                );
+                anchorBlockId_ = _params.anchorBlockId;
+            }
+
+            lastBlockTimestamp_ = _params.lastBlockTimestamp == 0
+                ? uint64(block.timestamp)
+                : _params.lastBlockTimestamp;
+
+            require(lastBlockTimestamp_ <= block.timestamp, TimestampTooLarge());
+
+            uint64 totalShift;
+            for (uint256 i; i < _params.blocks.length; ++i) {
+                totalShift += _params.blocks[i].timeShift;
+            }
+
+            require(lastBlockTimestamp_ >= totalShift, TimestampTooSmall());
```

Comment:
The `totalShift ` check can't be false, this [PR](https://github.com/taikoxyz/taiko-mono/pull/18784) trys to fix it.